File: ts/82.remove-duplicates-from-sorted-list-ii.ts
/*
 * @lc app=leetcode id=82 lang=typescript
 *
 * [82] Remove Duplicates from Sorted List II
 */

import ListNode from "./types/list-node";

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function deleteDuplicates(head: ListNode | null): ListNode | null {
  const dummy = new ListNode(0, head);

  let prev = dummy;

  while (head !== null) {
    if (head.next !== null && head.val === head.next.val) {
      while (head.next !== null && head.val === head.next.val) head = head.next;
      prev.next = head.next;
    } else {
      prev = prev.next;
    }

    head = head.next;
  }

  return dummy.next;
}
// @lc code=end

File: go/82.remove-duplicates-from-sorted-list-ii/82.remove-duplicates-from-sorted-list-ii.go
/*
 * @lc app=leetcode id=82 lang=golang
 *
 * [82] Remove Duplicates from Sorted List II
 */

package leetcode

import "github.com/maguroid/leetcode/go/structures"

type ListNode = structures.ListNode

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func deleteDuplicates(head *ListNode) *ListNode {
	dummy := &ListNode{Val: 0, Next: head}

	prev := dummy

	for head != nil {
		if isDuplicated(head) {
			for isDuplicated(head) {
				head = head.Next
			}
			prev.Next = head.Next
		} else {
			prev = prev.Next
		}

		head = head.Next
	}

	return dummy.Next
}

func isDuplicated(head *ListNode) bool {
	return head.Next != nil && head.Val == head.Next.Val
}

// @lc code=end
